icechunk/src/store.rs

```diff
@@ -428,31 +423,90 @@ impl Store {
         &self,
         prefix: &str,
     ) -> StoreResult<impl Stream<Item = StoreResult<ListDirItem>> + Send> {
-        // TODO: this is inefficient because it filters based on the prefix, instead of only
-        // generating items that could potentially match
-        // FIXME: this is not lazy, it goes through every chunk. This should be implemented using
-        // metadata only, and ignore the chunks, but we should decide on that based on Zarr3 spec
-        // evolution
-
-        let idx: usize = if prefix == "/" { 0 } else { prefix.len() };
-
-        let parents: HashSet<_> = self
-            .list_prefix(prefix)
-            .await?
-            .map_ok(move |s| {
-                // If the prefix is "/", get rid of it. This can happen when prefix is missing
-                // the trailing slash (as it does in zarr-python impl)
-                let rem = &s[idx..].trim_start_matches('/');
-                match rem.split_once('/') {
-                    Some((prefix, _)) => ListDirItem::Prefix(prefix.to_string()),
-                    None => ListDirItem::Key(rem.to_string()),
+        let session = Arc::clone(&self.session).read_owned().await;
+        let absolute_prefix =
+            if !prefix.starts_with("/") { &format!("/{}", prefix) } else { prefix };
+        let prefix = prefix.trim_end_matches("/");
+
+        #[allow(clippy::expect_used)]
+        let path =
+            Path::try_from(absolute_prefix).expect("could not create path from prefix");
+        let results = match session.get_node(&path).await {
+            Ok(NodeSnapshot {
+                id: node_id,
+                node_data: NodeData::Array(.., manifests),
+                ..
+            }) => {
+                // if this is an array we know what to yield
+                let mut res = vec![ListDirItem::Key("zarr.json".to_string())];
+                if session.array_has_chunks_set(&node_id) || !manifests.is_empty() {
```

Comment:
hmm I don't think this condition is precise enough. It could have chunks in the manifests but all deleted by the session. I think we probably want to call some existing function that lists all chunks for the array and just verify if there are any.

Reply:
yeah I was trying to be too clever about not downloading a manifest. That can be a future optimization.

Reply:
Ahh I forgot about that! I think this is worth exploring because it can be a big optimization. Maybe if there are manifests and no chunk deletes....? We would have to think about edge cases ... but I like this idea.

Reply:
OK I tried again.

Reply:
Now I've added a couple more deletion rules to the stateful tests in addition to a list_dir rule to gain more confidence.

Reply:
OK I think the one edge-case not addressed is when the user calls `delete_dir("path/to/array/c")`. This is not a deleted array, and we can't detect this without checking that every chunk of the array was deleted or never written in the first place :/